### PR TITLE
Use optimized atomicAllInc

### PR DESCRIPTION
**Pull request by Flamefire**
This uses atomicAllInc instead of atomicAdd(..., 1) which is optimized for causing less collisions

### CODE REVIEW

File: src/picongpu/include/plugins/kernel/CopySpeciesGlobal2Local.kernel [hunk 1]
```diff
@@ -113,7 +114,7 @@ __global__ void copySpeciesGlobal2Local(T_CounterBox counter, T_DestBox destBox,
         if (masterIdx == linearThreadIdx)
         {
             /* counter[2] -> number of used frames */
-            atomicAdd(&(counter[2]), 1u);
+            nvidia::atomicAllInc(&(counter[2]));
```
Comment (ax3l) [inline src/picongpu/include/plugins/kernel/CopySpeciesGlobal2Local.kernel]: compile suite says:
```
error: argument of type "uint32_t *" is incompatible with parameter of type "int *"
```

File: src/picongpu/include/plugins/kernel/CopySpeciesGlobal2Local.kernel [hunk 2]
```diff
@@ -132,7 +133,7 @@ __global__ void copySpeciesGlobal2Local(T_CounterBox counter, T_DestBox destBox,
         /* counter[1] -> number of loaded particles
          * this counter is evaluated on host side
          * (check that loaded particles by this kernel == loaded particles from HDF5 file)*/
-        atomicAdd(&(counter[1]), 1u);
+        nvidia::atomicAllInc(&(counter[1]));
```
Comment (ax3l) [inline src/picongpu/include/plugins/kernel/CopySpeciesGlobal2Local.kernel]: same problem here
```
error: argument of type "uint32_t *" is incompatible with parameter of type "int *"
```

Comment (Flamefire) [inline src/picongpu/include/plugins/kernel/CopySpeciesGlobal2Local.kernel]: @ax3l Fixed by extending the function to handle those types

Comment (ax3l) [inline src/picongpu/include/plugins/kernel/CopySpeciesGlobal2Local.kernel]: well done, thx.

**Comment by ax3l** [comment]
I'll let the rest of the review to @psychocoderHPC since he is way more fluent in those methods.

File: src/libPMacc/include/nvidia/atomic.hpp [hunk 3]
```diff
@@ -51,22 +52,28 @@ namespace nvidia
...
-    const int leader = __ffs(mask) - 1;
-    int restult;
-    const int lanId = getLaneId();
-    if (lanId == leader)
-        restult = atomicAdd(ptr, __popc(mask));
-    restult = warpBroadcast(restult, leader);
-    /* each thread computes its own value */
-    return restult + __popc(mask & ((1 << lanId) - 1));
+    // warpBroadcast is only defined for int (unsigned int is also ok through conversion)
+    if(boost::is_same<T, int>::value || boost::is_same<T, unsigned>::value)
```
Comment (ax3l) [inline src/libPMacc/include/nvidia/atomic.hpp]: brr, can't you rather write a specialization of the function for those types ? ;)

File: src/libPMacc/include/nvidia/atomic.hpp [hunk 4]
```diff
@@ -51,22 +52,28 @@ namespace nvidia
  * nvidia parallel forall example
  * http://devblogs.nvidia.com/parallelforall/cuda-pro-tip-optimized-filtering-warp-aggregated-atomics/
  */
+template<typename T>
 DINLINE
-int atomicAllInc(int *ptr)
+T atomicAllInc(T *ptr)
 {
 #if (__CUDA_ARCH__ >= 300)
```
Comment (ax3l) [inline src/libPMacc/include/nvidia/atomic.hpp]: even that could become a template parameter and could be used for a specialization

File: src/picongpu/include/plugins/kernel/CopySpeciesGlobal2Local.kernel (hunk 1 above)
Comment (ax3l) [inline src/picongpu/include/plugins/kernel/CopySpeciesGlobal2Local.kernel]: fixes with implementation

**Comment by Flamefire** [comment]
This requires #1078

**Comment by Flamefire** [comment]
Rebased

**Pull request merged**